usr/local/qterm/options.3
.\"
.\" Copyright (c) 1990 Michael A. Cooper.
.\" This software may be freely distributed provided it is not sold for 
.\" profit and the author is credited appropriately.
.\"
.\" $Header: /src/common/usc/lib/libgen/RCS/options.3,v 1.4 1991/02/22 02:30:06 mcooper Exp $
.\"
.TH PARSEOPTIONS 3 "30 October 1990"
.ds ]W USC-UCS
.SH NAME
ParseOptions, UsageOptions, HelpOptions, Num_Opts \- Parse command line options
.SH SYNOPSIS
.LP
.nf
.ft B
#include "/usr/local/include/options.h"
.ft
.fi
.LP
.nf
.ft B
int ParseOptions(options, num_options, argc, argv)
OptionDescRec *options;
int num_options;
int argc;
char **argv;
.ft
.fi
.LP
.nf
.ft B
void UsageOptions(options, num_options, badoption)
OptionDescRec *options;
int num_options;
char *badoption;
.ft
.fi
.LP
.nf
.ft B
void HelpOptions(options, num_options, message)
OptionDescRec *options;
int num_options;
char **message;
.ft
.fi
.LP
.nf
.ft B
int Num_Opts(options)
OptionDescRec *options;
.ft
.fi
.LP
.nf
.ft B
extern char *OptionChars;
extern char *ProgramName;
.ft
.fi
.SH DESCRIPTION
.LP
.BR ParseOptions(\|)
parses a given set of options found in
.B argv.
The
.B argc
parameter is the count of the number of string pointers
in 
.B argv.
Both 
.B argc
and
.B argv
are typically passed directly from a
.B main(\|)
function.
The
.B argv
parameter should contain an array of strings that
need to be parsed.
.B ParseOptions(\|)
returns the number of entries in
.B argv
that were successfully parsed or -1 upon error.
.LP
The
.B options
structure should contain a valid description list of options.
The type
.B OptionDescRec
is defined as the following in the
.B "options.h"
header file:
.RS
.LP
.nf
.ft B
typedef struct {
    char      *option;     /* Option string in argv */
    int       flags;       /* Flag bits */
    int       (*cvtarg)(); /* Function to convert argument */
    caddr_t   valp;        /* Variable to set */
    caddr_t   value;       /* Default value to provide */
    char      *usage;      /* Usage message */
    char      *desc;       /* Description message */
} OptionDescRec, *OptionDescList;
.ft R
.fi
.RE
.LP
The order of
.I options
is important because
the first partial match found in
.I options 
is used.
This allows abbreviations (except if the option is a
.I StickArg
[see below]).
For instance, a user may specify only "\-n" for "\-number" provided
that "\-n" is unique to
.I options
or that "\-number" is placed before any other "\-n*" options in
.I options.
.LP
The
.I option
member of
.B OptionDescRec
is the string name of the option.
This is typically something like
.RS
.ft B
.nf
.sp
"\-c"
"+c"
"\-file"
.fi
.sp
.ft
.RE
The first character of 
.I option
is special.  It must be one of the characters know to be the
start of an option.
The default list of starting option characters is "\-+".
This indicates that an option can start with either a "\-" or
a "+".  This list of characters may be changed by setting
the variable
.B OptionChars
to point to a string of custom starting option characters.
.LP
The
.I flags
member is used to set bits to describe how an option
is to be interpreted.
Valid flags are defined in the
.I "options.h"
header file:
.RS
.IP NoArg
No argument for this option.  
Use the value in 
.I OptionDescRec.value 
to set the value in the
.I valp
member of 
.B OptionDescRec.
.IP IsArg
Value is the option string itself.
.IP SepArg
Value is in next argument in argv.
.IP StickyArg
Value is the characters immediately following 
the option.
.IP SkipArg
Ignore this option and the next argument in argv.
.IP SkipLine
Ignore this option and the rest of argv.
.IP SkipNArgs
Ignore this option and the next 
.I OptionDescRes.value 
arguments in argv.
.IP ArgHidden
Don't show this option in usage or help messages.
.RE
.LP
The next member of
.B OptionDescRec
is
.I cvtarg.
This should be a pointer to a function that knows how to
convert the given argument into the correct type.
The predefined functions are as follows:
.RS
.IP OptBool
Converts a boolean.
.IP OptInt
Converts an integer.
.IP OptShort
Converts a short.
.IP OptLong
Converts a long.
.IP OptStr
Converts a string.
.RE
.LP
The
.I valp
member should be a pointer
to the variable that needs to be set.
.I valp
should accept whatever type
.I cvtarg
is expected to return.
.LP
The
.I value
member should contain a default value to
be used if no value is given for an option or
this type of option does not require an argument
(according to the 
.I flags
bits).
If 
.I value 
is
.B NULL
then an argument for this option
is optional.
.LP
.I usage
is used to build usage and help messages.
It should be a string containing a description of any arguments
that may be used for this option.
The option string itself should not be a part of 
.I usage.
The 
.B UsageOptions(\|)
and 
.B HelpOptions(\|)
functions combine the
.I option
field with
.I usage
and interpret the
.I flags
bits to build a usage string.
If this field is 
.B NULL,
then just the
.I option
field itself is used for usage and help messages.
.LP
The
.I desc
member is used to build a help message for this option.
This should be a string containing a brief description on what this
option does.
.LP
The
.B num_options
parameter should be the number of 
.B OptionDescRec's
found in 
.B options.
The function
.BR Num_Opts(\|)
will return the number of 
.B OptionDescRec's.
.LP
The
.B UsageOptions(\|)
function
prints a usage message.
If
.I badoption
is not 
.B NULL, 
then an initial message is displayed indicating that 
.I badoption
is not a valid option.
.LP
The
.B HelpOptions(\|)
function
prints a nicely formatted message describing all options.
If
.I message
is not 
.B NULL
it is taken to be a message that is displayed in the output of
a "\-help" option.
.SH EXAMPLE
.LP
Here is an example program:
.nf
.sp
.ft B
#include "options.h"

char *filename = NULL;
int number = \-1;
int foo = \-1;
int I = \-1;
long L = \-1;
short S = \-1;

OptionDescRec opts[] = {
    {"\-foo",	NoArg,		OptBool, (caddr_t) &foo,	"0",
     (char *)NULL,	"Disable foo bar"},
    {"+foo",	NoArg,		OptBool, (caddr_t) &foo,       	"1",
     (char *)NULL,	"Enable foo bar"},
    {"\-I",	StickyArg,	OptInt, (caddr_t) &I,		(caddr_t) NULL,
     (char *)NULL,	"Set value of I"},
    {"\-L",	StickyArg,	OptLong, (caddr_t) &L,		(caddr_t) NULL,
     (char *)NULL,	"Set value of L"},
    {"\-S",	SepArg,		OptShort, (caddr_t) &S,		(caddr_t) NULL,
     (char *)NULL,	"Set value of S"},
    {"\-C",	StickyArg,	OptStr, (caddr_t) &filename,	(caddr_t) NULL,
     (char *)NULL,	"Alternate file to use"},
    {"\-number",	SepArg, 	OptInt, (caddr_t) &number,	"66",
     "interval",	NULL},
    {"\-file",	SepArg, 	OptStr, (caddr_t) &filename,	(caddr_t) NULL,
     "filename",	"Specify alternate file to use"},
};

main(argc, argv)
     int argc;
     char **argv;
{
    int c;

    c = ParseOptions(opts, Num_Opts(opts), argc, argv);
    printf("Count = %d of %d\n", c, argc);
}
.ft
.fi
.SH "RETURN VALUES"
.B ParseOptions(\|)
returns the number of arguments parsed or -1 upon error.
.SH NOTES
.LP
The
.I \-help
option is automatically built into 
.B ParseOptions(\|).
.LP
All error messages are sent to 
.B stderr.
.LP
An option may be both 
.I StickyArg
and
.I SepArg.
If both are set for one option, preference is given to
.I SepArg
parsing.
Also, no appreviations are allowed.
.SH AUTHOR
Michael A. Cooper, 
.br
University Computing Services, 
.br
University of Southern California.


usr/local/qterm/options.c
/*
 * Copyright (c) 1990 Michael A. Cooper.
 * This software may be freely distributed provided it is not sold for 
 * profit and the author is credited appropriately.
 */

#if	!defined(lint) && defined(DOSCCS)
static char *RCSid = "$Header: options.c,v 1.13.1 96/3/23 18:13:28 sms Exp $";
#endif

/*
 * $Log:	options.c,v $
 * Revision 1.13  90/12/15  18:13:28  mcooper
 * Add copywrite notice.
 * 
 * Revision 1.12  90/12/15  17:51:46  mcooper
 * Add #ifdef HAS_VARARGS around include for <varargs>.
 * 
 * Revision 1.11  90/11/13  16:39:28  mcooper
 * Add #ifdef HAS_VARARGS for systems without
 * varargs.
 * 
 * Revision 1.10  90/11/13  15:28:01  mcooper
 * - Add OptBool cvtarg routine.
 * - Print default values in HelpOptions() 
 *   when appropriate.
 * 
 * Revision 1.9  90/11/13  15:19:00  mcooper
 * Added supported for options being both
 * SepArg and StickyArg.
 * 
 * Revision 1.8  90/10/30  21:02:31  mcooper
 * Need to exit() if -help is specified.
 * 
 * Revision 1.7  90/10/30  20:24:33  mcooper
 * Fixed bug in UsageString().
 * 
 * Revision 1.6  90/10/30  19:53:05  mcooper
 * Cleaned up some NeXT cc and lint stuff.
 * 
 * Revision 1.5  90/10/30  19:45:31  mcooper
 * Remove unneeded paramter to HelpOptions().
 * 
 * Revision 1.4  90/10/29  14:47:42  mcooper
 * Added real function UsageString() to
 * handle formating usage option strings.
 * 
 * Revision 1.3  90/10/29  14:17:00  mcooper
 * Allow options to be abbreviated 
 * (for all non StickArg options).
 * 
 * Revision 1.2  90/10/26  15:56:11  mcooper
 * - Fix bug in SepArg code that ate arguments.
 * - Cleanup help message.
 * - Add ArgHidden code.
 * 
 * Revision 1.1  90/10/26  14:42:51  mcooper
 * Initial revision
 * 
 */

/*
 * Functions to parse options.
 */

#include "options.h"
#ifdef HAS_VARARGS
#include <varargs.h>
#endif

char *OptionChars = "-+";	/* Default option switching characters */
char *ProgramName = NULL;	/* Name of this program */
char *UsageString();
static int isopt();
static int suppress_help_msg = 0;
char *strcat();

/*
 * ParseOptions - Parse options found in argv using "options".
 *		  Returns the number of options parsed if there
 *		  were no errors.  Returns -1 if an error occurs.
 */
int ParseOptions(options, num_options, argc, argv)
     OptionDescRec *options;
     int num_options;
     int argc;
     char **argv;
{
    OptionDescRec *opt;
    register int x;
    char *p;

    if (ProgramName == NULL)
	ProgramName = argv[0];

#ifdef OPTION_DEBUG
    (void) printf("Option list is:\n");
    for (x = 0; x < num_options; ++x) {
	opt = &options[x];
	(void) printf("%s\n", opt->option);
    }

    (void) printf("Arguments (%d):", argc);
    for (x = 0; x < argc; ++x) {
	(void) printf(" %s", argv[x]);
    }
    (void) printf("\n");
#endif /* OPTION_DEBUG */

    for (x = 1; x < argc; ++x) {
	if (strcmp(HELPSTR, argv[x]) == 0) {
	    HelpOptions(options, num_options, (char **)NULL);
	    exit(0);
	}

	opt = FindOption(options, num_options, argv[x]);
	if (opt == NULL) {
	    if (isopt(argv[x])) { /* this was suppose to be an option */
		UsageOptions(options, num_options, argv[x]);
		return(-1);
	    } else { /* must be end of options */
		break;
	    }
	}

	if (opt->flags & NoArg) {
	    if (!(*opt->cvtarg)(opt, opt->value, FALSE)) {
		UsageOptions(options, num_options, opt->option);
		return(-1);
	    }
	} else if (opt->flags & IsArg) {
	    if (!(*opt->cvtarg)(opt, opt->option, FALSE)) {
		UsageOptions(options, num_options, opt->option);
		return(-1);
	    }
	} else if ((opt->flags & StickyArg) && (opt->flags & SepArg)) {
	    p = (char *) &argv[x][strlen(opt->option)];
	    if (!*p) {		/*** SepArg ***/
		if (x + 1 >= argc || isopt(argv[x+1])) {
		    if (opt->value == (caddr_t) NULL) {
			UserError("%s: Option requires an argument.", argv[x]);
			UsageOptions(options, num_options, opt->option);
			return(-1);
		    }
		    p = opt->value;
		} else {
		    p = argv[++x];
		}
	    }
	    if (!(*opt->cvtarg)(opt, p, TRUE)) {
		UsageOptions(options, num_options, opt->option);
		return(-1);
	    }
	} else if (opt->flags & StickyArg) {
	    p = (char *) &argv[x][strlen(opt->option)];
	    if (!*p) {
		if (opt->value == (caddr_t) NULL) {
		    UserError("%s: Option requires an argument.", argv[x]);
		    UsageOptions(options, num_options, opt->option);
		    return(-1);
		} else {
		    p = opt->value;
		}
	    }
	    if (!(*opt->cvtarg)(opt, p, TRUE)) {
		UsageOptions(options, num_options, opt->option);
		return(-1);
	    }
	} else if (opt->flags & SepArg) {
	    if (x + 1 >= argc || isopt(argv[x+1])) {
		if (opt->value == (caddr_t) NULL) {
		    UserError("%s: Option requires an argument.", argv[x]);
		    UsageOptions(options, num_options, opt->option);
		    return(-1);
		} else {
		    p = opt->value;
		}
	    } else {
		p = argv[++x];
	    }
	    if (!(*opt->cvtarg)(opt, p, TRUE)) {
		UsageOptions(options, num_options, opt->option);
		return(-1);
	    }
	} else if (opt->flags & SkipArg) {
	    x += 2;
	} else if (opt->flags & SkipLine) {
	    return(x);
	} else if (opt->flags & SkipNArgs) {
	    if (opt->value) {
		x += atoi(opt->value);
	    } else {
		UserError("Internal Error: No 'value' set for SkipNArgs.");
		return(-1);
	    }
	} else {
	    UserError("Internal Error: Unknown argument type for option '%s'.",
		     opt->option);
	    return(-1);
	}
    }

    return(x);
}

/*
 * FindOption - Find "option" in "options".  Returns NULL if not found.
 */
OptionDescRec *FindOption(options, num_options, option)
     OptionDescRec *options;
     int num_options;
     char *option;
{
    OptionDescRec *opt;
    register int x;

    for (x = 0; x < num_options; ++x) {
	opt = &options[x];
	if (opt->flags & StickyArg) {
	    if (strncmp(option, opt->option, strlen(opt->option)) == 0)
		return(opt);
	} else {
	    if (strncmp(option, opt->option, strlen(option)) == 0)
		return(opt);
	}
    }

    return(NULL);
}

/*
 * isopt - Is "str" an option string?  Compare first char of str against
 *	   list of option switch characters.  Returns TRUE if it is an option.
 */
static int isopt(str)
     char *str;
{
    register char *p;

    for (p = OptionChars; p && *p; *++p) {
	if (*str == *p) {
	    return(TRUE);
	}
    }

    return(FALSE);
}

/*
 * UsageOptions - Print a usage message based on "options".
 */
void UsageOptions(options, num_opts, badOption)
     OptionDescRec *options;
     int num_opts;
     char *badOption;
{
    OptionDescRec *opt;
    char *optstr;
    register int x;
    int col, len;

    if (badOption) 
	(void) fprintf (stderr, "%s:  bad command line option \"%s\"\r\n\n",
			ProgramName, badOption);

    (void) fprintf (stderr, "usage:  %s", ProgramName);
    col = 8 + strlen(ProgramName);
    for (x = 0; x < num_opts; x++) {
	opt = &options[x];
	if (opt->flags & ArgHidden)
	    continue;
	optstr = UsageString(opt);
	len = strlen(optstr) + 3;	/* space [ string ] */
	if (col + len > 79) {
	    (void) fprintf (stderr, "\r\n   ");  /* 3 spaces */
	    col = 3;
	}
	(void) fprintf (stderr, " [%s]", optstr);
	col += len;
    }

    if (suppress_help_msg)
	(void) fprintf(stderr, "\r\n\n");
    else
	(void) fprintf(stderr, 
		       "\r\n\nType \"%s %s\" for a full description.\r\n\n",
		       ProgramName, HELPSTR);
}

/*
 * HelpOptions - Print a nice help/usage message based on options.
 */
void HelpOptions(options, num_opts, message)
     OptionDescRec *options;
     int num_opts;
     char **message;
{
    OptionDescRec *opt;
    register int x;
    char **cpp;

    suppress_help_msg = 1;
    UsageOptions(options, num_opts, (char *)NULL);
    suppress_help_msg = 0;

    (void) fprintf (stderr, "where options include:\n");
    for (x = 0; x < num_opts; x++) {
	opt = &options[x];
	if (opt->flags & ArgHidden)
	    continue;
	(void) fprintf (stderr, "    %-28s %s\n", UsageString(opt), 
		 (opt->desc) ? opt->desc : "");
	if (opt->value && opt->cvtarg != OptBool)
	    (void) fprintf (stderr, "    %-28s [ Default value is %s ]\n", 
			    "", opt->value);
    }

    if (message) {
	(void) putc ('\n', stderr);
	for (cpp = message; *cpp; cpp++) {
	    (void) fputs (*cpp, stderr);
	    (void) putc ('\n', stderr);
	}
	(void) putc ('\n', stderr);
    }
}

/*
 * UserError - Print a user error.
 */
#ifdef HAS_VARARGS
void UserError(va_alist)
     va_dcl
{
    va_list args;
    char *fmt;

    va_start(args);
    if (ProgramName)
	(void) fprintf(stderr, "%s: ", ProgramName);
    fmt = (char *) va_arg(args, char *);
    (void) vfprintf(stderr, fmt, args);
    va_end(args);
    (void) fprintf(stderr, "\n");
}
#else
void UserError(fmt, a1, a2, a3, a4, a5, a6)
     char *fmt;
{
    if (ProgramName)
	(void) fprintf(stderr, "%s: ", ProgramName);
    (void) fprintf(stderr, fmt, a1, a2, a3, a4, a5, a6);
    (void) fprintf(stderr, "\n");
}
#endif

OptBool(opt, value, docopy)
     OptionDescRec *opt;
     caddr_t value;
     int docopy; /*ARGSUSED*/
{
    char *vpp;

    *(int *) opt->valp = (int) strtol(value, &vpp, 0);
    if (*vpp) {
	UserError("Invalid integer argument for '%s'.", opt->option);
	return(FALSE);
    } else {
	return(TRUE);
    }
}

OptInt(opt, value, docopy)
     OptionDescRec *opt;
     caddr_t value;
     int docopy; /*ARGSUSED*/
{
    char *vpp;

    *(int *) opt->valp = (int) strtol(value, &vpp, 0);
    if (*vpp) {
	UserError("Invalid integer argument for '%s'.", opt->option);
	return(FALSE);
    } else {
	return(TRUE);
    }
}

OptShort(opt, value, docopy)
     OptionDescRec *opt;
     caddr_t value;
     int docopy; /*ARGSUSED*/
{
    char *vpp;

    *(short *) opt->valp = (short) strtol(value, &vpp, 0);
    if (*vpp) {
	UserError("Invalid integer argument for '%s'.", opt->option);
	return(FALSE);
    } else {
	return(TRUE);
    }
}

OptLong(opt, value, docopy)
     OptionDescRec *opt;
     caddr_t value;
     int docopy; /*ARGSUSED*/
{
    char *vpp;

    *(long *) opt->valp = (long) strtol(value, &vpp, 0);
    if (*vpp) {
	UserError("Invalid integer argument for '%s'.", opt->option);
	return(FALSE);
    } else {
	return(TRUE);
    }
}

OptStr(opt, value, docopy)
     OptionDescRec *opt;
     caddr_t value;
     int docopy;
{
    char *p;

    if (docopy) {
	if ((p = (char *) malloc((unsigned)strlen(value)+1)) == NULL) {
	    UserError("Cannot malloc memory.");
	    return(FALSE);
	}
	(void) strcpy(p, value);
    } else {
	p = value;
    }

    *(char **) opt->valp = p;

    return(TRUE);
}

static char *UsageString(opt)
     OptionDescRec *opt;
{
    static char buf[BUFSIZ], buf2[BUFSIZ];

    (void) sprintf(buf, opt->option);
    (void) strcpy(buf2, "");
    if (opt->usage) {
	(void) sprintf(buf2, "%s%s%s%s",
		       ((opt->flags & StickyArg) && 
			!((opt->flags & StickyArg) && (opt->flags & SepArg))) 
		       ? "" : " ",
		       (opt->value) ? "[" : "",
		       opt->usage,
		       (opt->value) ? "]" : ""
		       );
    }
    (void) strcat(buf, buf2);

    return(buf);
}
